classes/local/form/preferences_form.php

```diff
@@ -81,34 +85,42 @@ public function define_configured_factors($mform) {
             'centeralign',
             'centeralign',
             'centeralign',
+            'centeralign',
+            'centeralign',
         );
         $table->data  = array();
 
         $factors = \tool_mfa\plugininfo\factor::get_enabled_factors();
 
         foreach ($factors as $factor) {
 
-            $userfactors = $factor->get_all_user_factors();
+            $userfactors = $factor->get_enabled_user_factors();
 
             foreach ($userfactors as $userfactor) {
                 $url = "action.php?sesskey=" . sesskey();
                 $edit = "<a href=\"action.php?sesskey=".sesskey()
                     ."&amp;action=edit&amp;factor=$factor->name&amp;factorid=$userfactor->id\">$headers->edit</a>";
-                $revoke = "<a href=\"action.php?sesskey=".sesskey()
-                    ."&amp;action=revoke&amp;factor=$factor->name&amp;factorid=$userfactor->id\">$headers->revoke</a>";
+                if ($factor->has_delete()) {
+                    $revokeparams = "action=revoke&amp;factor=$factor->name&amp;factorid=$userfactor->id";
```

Comment:
Please don't construct urls like this manually. Use moodle_url() so they are escaped properly.

The same goes for links, please use html_writer:

html_writer::link($logouturl, get_string('logout'));